core/src/test/java/com/linecorp/armeria/internal/client/DefaultClientRequestContextTest.java

```diff
@@ -214,19 +215,109 @@ void nestedContextCaptors() {
         assertThat(ClientThreadLocalState.get()).isNull();
     }
 
+    @Test
+    void testAuthorityOverridden() {
+        final HttpRequest request1 = HttpRequest.of(RequestHeaders.of(
+                HttpMethod.POST, "/foo",
+                HttpHeaderNames.SCHEME, "http"));
+        final DefaultClientRequestContext ctx = newContext(ClientOptions.of(), request1);
+        assertThat(ctx.authority()).isNull();
+        assertThat(ctx.uri().toString()).isEqualTo("http:/foo");
+        assertThat(ctx.uri()).hasScheme("http").hasAuthority(null).hasPath("/foo");
+
+        ctx.init(Endpoint.of("endpoint.com", 8080));
+        assertThat(ctx.authority()).isEqualTo("endpoint.com:8080");
+        assertThat(ctx.uri().toString()).isEqualTo("http://endpoint.com:8080/foo");
+
+        final HttpRequest request2 = HttpRequest.of(RequestHeaders.of(
+                HttpMethod.POST, "/bar",
+                HttpHeaderNames.SCHEME, "http",
+                HttpHeaderNames.AUTHORITY, "request.com"));
+        ctx.updateRequest(request2);
+        assertThat(ctx.authority()).isEqualTo("request.com");
+        assertThat(ctx.uri().toString()).isEqualTo("http://request.com/bar");
+
+        ctx.addAdditionalRequestHeader(HttpHeaderNames.AUTHORITY, "additional.com");
+        assertThat(ctx.authority()).isEqualTo("additional.com");
+        assertThat(ctx.uri().toString()).isEqualTo("http://additional.com/bar");
+    }
+
+    @Test
+    void testDefaultAuthorityOverridesInternal() {
+        final HttpRequest request1 = HttpRequest.of(RequestHeaders.of(
+                HttpMethod.POST, "/foo",
+                HttpHeaderNames.SCHEME, "http"));
+        final ClientOptions clientOptions = ClientOptions.builder()
+                                                         .addHeader(HttpHeaderNames.AUTHORITY, "default.com")
+                                                         .build();
+        final DefaultClientRequestContext ctx = newContext(clientOptions, request1);
+        ctx.init(Endpoint.of("example.com", 8080));
+        assertThat(ctx.authority()).isEqualTo("default.com");
+        assertThat(ctx.uri().toString()).isEqualTo("http://default.com/foo");
+    }
+
+    @Test
+    void requestUpdateOnlyUpdatesPathAndQuery() {
+        final DefaultClientRequestContext ctx = newContext();
+        assertThat(ctx.authority()).isEqualTo("example.com:8080");
+        assertThat(ctx.uri().toString()).isEqualTo("http://example.com:8080/foo");
+
+        final HttpRequest request = HttpRequest.of(RequestHeaders.of(
+                HttpMethod.POST, "https://path.com/a/b/c",
+                HttpHeaderNames.SCHEME, "http",
+                HttpHeaderNames.AUTHORITY, "request.com"));
+        ctx.updateRequest(request);
+        // the scheme and authority in path is ignored
+        assertThat(ctx.sessionProtocol()).isEqualTo(SessionProtocol.H2C);
+        assertThat(ctx.authority()).isEqualTo("request.com");
```

Comment:
I understood your question as: why do we ignore the scheme and authority when updating the request? (which seems different from `DefaultWebClient` which uses the scheme, authority in a request)

Originally I didn't want to modify the endpoint that was acquired because I thought it was easy to misuse this api.

Having said this, I agree that just updating the scheme and endpoint is easier to reason about and yields consistent behavior. Let me try updating sessionProtocol, endpoint and see if anything breaks